### PR TITLE
Adding namespace handler to add project id label

### PR DESCRIPTION
**Pull request by leodotcloud**

### CODE REVIEW

File: pkg/controllers/user/nslabels/labels.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		cluster.Core.Namespaces("").ObjectClient(),
+	}
+	cluster.Core.Namespaces("").AddHandler("namespaceHandler", nsh.Sync)
+}
+
+func (nsh *namespaceHandler) Sync(key string, ns *corev1.Namespace) error {
+	if ns == nil {
+		return nil
+	}
+	logrus.Infof("namespaceHandler Sync key=%v, ns=%+v", key, *ns)
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: This is going to log too much.  You should really only log if you actually perform an action.  Sync handlers are called for ever change and then also on an interval.

File: pkg/controllers/user/nslabels/labels.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	metaAccessor, err := meta.Accessor(ns)
+	if err != nil {
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: You don't need an accessor, annotations are available as a field, `ns.Annotations["foo"]`

File: pkg/controllers/user/nslabels/labels.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+const (
+	projectIDField = "field.cattle.io/projectId"
+)
+
+type namespaceHandler struct {
+	nsLister typescorev1.NamespaceLister
+	nsClient *clientbase.ObjectClient
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: use the typed interface for the client `v3.NamespaceInterface`

File: pkg/controllers/user/nslabels/labels.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func Register(cluster *config.UserContext) {
+	logrus.Infof("Registering namespaceHandler for adding labels ")
+	nsh := &namespaceHandler{
+		cluster.Management.Core.Namespaces("").Controller().Lister(),
+		cluster.Core.Namespaces("").ObjectClient(),
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: drop `.ObjectClient()`

File: pkg/controllers/user/nslabels/labels.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (nsh *namespaceHandler) addProjectIDLabelToNamespace(ns *corev1.Namespace, projectID string) error {
+	var err error
+	nscopy := ns.DeepCopy()
+	if nscopy.Labels == nil {
+		nscopy.Labels = map[string]string{}
+	}
+	nscopy.Labels[projectIDField] = projectID
+	if !reflect.DeepEqual(ns, nscopy) {
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: You don't need to use reflection, just 
```golang
if ns.Labels[projectIDField] != projectID {
// deepcopy, set label, call Update
}
```

File: pkg/controllers/user/nslabels/labels.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+const (
+	projectIDField = "field.cattle.io/projectId"
+)
+
+type namespaceHandler struct {
+	nsLister typescorev1.NamespaceLister
```
Comment (ibuildthecloud) [inline pkg/controllers/user/nslabels/labels.go]: Not used, you can delete this.

File: pkg/controllers/user/nslabels/labels.go (hunk 1 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: Ok, will move it to debug

File: pkg/controllers/user/nslabels/labels.go (hunk 2 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: 👍

File: pkg/controllers/user/nslabels/labels.go (hunk 3 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: ok

File: pkg/controllers/user/nslabels/labels.go (hunk 4 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: ok

File: pkg/controllers/user/nslabels/labels.go (hunk 5 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: thank you, will change.

File: pkg/controllers/user/nslabels/labels.go (hunk 6 above)
Comment (leodotcloud) [reply pkg/controllers/user/nslabels/labels.go]: good catch

**Pull request merged**